### PR TITLE
ESP8266 - specifying station BSSID to connect to

### PR DESCRIPTION
**Pull request by eimiss**
Hello,
[noob alert]
Seems like Simba doesn't have an option to specify WiFi station BSSID to connect to. This is important when connecting to multi-AP WiFi network because SDK is too dumb to figure out which AP to connect to on its own.
Espressiv SDK does support this via the station_config struct. I'm wondering whether it's planned to add this into Simba, or am I just missing something?

### CODE REVIEW

**Comment by eimiss** [comment]
I tried to add the BSSID functionality. Decided to utilize the existing arg_p to transfer the BSSID into the esp_wifi_station_port_init function. Not sure if this implementation is good enough, awaiting critique :smile:

File: examples/esp_wifi/station/main.c [hunk 1]
```diff
@@ -43,7 +43,9 @@ int main()
     inet_aton("255.255.255.0", &info.netmask);
     inet_aton("192.168.0.1", &info.gateway);
 
-    if (esp_wifi_station_init("Qvist2", "maxierik", &info) != 0) {
+    if (esp_wifi_station_init("Qvist2", "maxierik",
```
Comment (eerimoq) [inline examples/esp_wifi/station/main.c]: One argument per line.

File: src/drivers/ports/esp/esp_wifi_port/station.i [hunk 2]
```diff
@@ -60,6 +61,11 @@ static int esp_wifi_station_port_init(const char *ssid_p,
         config.password[0] = 0;
     }
 
+    if (bssid_p != NULL) {
+        config.bssid_set = 1;
+        memcpy(config.bssid, bssid_p, 6);
```
Comment (eerimoq) [inline src/drivers/ports/esp/esp_wifi_port/station.i]: `memcpy(config.bssid, bssid_p, sizeof(config.bssid));`

**Comment by eimiss** [comment]
Well if I'm not mistaken it's 100% copy & paste for esp32's station.i

**Comment by eerimoq** [comment]
Both ports compiles just fine after your changes. :-)

I noticed that the config variable is not cleared at the beginning of the init function, so there are fields that are only implicitly initialized/cleared by the startup code.

Can you add a `memset(&config, 0, sizeof(config))` or initialize all fields individually in both ports as well? Should be easy to test since you already have the hardware setup ready for the ESP8266. I can test ESP32 later.

**Comment by eimiss** [comment]
Ha, I saw that too but thought lack of init was deliberate, also my ESP8266 works just fine that way. Could it be that somewhere in the chain those fields actually get implicitly initialized?
```C
\\ arg_p = NULL of course
std_printf(FSTR("%d\n\r%x:%x:%x:%x:%x:%x\n\r"), config.bssid_set,
                    config.bssid[0], config.bssid[1], config.bssid[2],
                    config.bssid[3], config.bssid[4], config.bssid[5]);
```
Outputs:
```
0
0:0:0:0:0:0
```
I guess I'll add ```static struct station_config config = {0};``` anyways, to be explicit, with your approval.

**Comment by eerimoq** [comment]
Sounds good, but more importantly initialize all fields in the config variable in the `esp_wifi_port_station_init()` function. Otherwise the bssid flag cannot be unset.

**Comment by eimiss** [comment]
I understand and hate to argue with you, but doing:
```C
static struct station_config config = {0};
```
makes
```C
config.bssid_set = 0;
```
redundant. Doesn't it? It's in accordance with C spec and looks clean.

**Comment by eerimoq** [comment]
Hmm, but what if `esp_wifi_station_init()` is first called with `bssid_p != NULL`, and then called again with `bssid_p == NULL`? Wouldn't the second call also configure the WiFi with bssid set, which is unexpected since `bssid_p` was set to `NULL`?

**Comment by eimiss** [comment]
Yes you are absolutely right. Now I truly _understand_ :D

**Comment by eerimoq** [comment]
=P

**Pull request merged**